From b7780dab90e8da302bc7d0d1b39b538b822017e8 Mon Sep 17 00:00:00 2001
From: Corey Minyard <cminyard@mvista.com>
Date: Thu, 5 Apr 2018 16:44:12 -0500
Subject: [PATCH] ipmi: Add shutdown functions for users and interfaces
Git-commit: b7780dab90e8da302bc7d0d1b39b538b822017e8
Patch-mainline: v4.18-rc1
References: FATE#326156

Since things that IPMI uses can be hot-swapped, the users and
interfaces really need to be able to handle this.

Add the functions so the users and interfaces can implement
them, the actual function will be added after everything is
ready.

Signed-off-by: Corey Minyard <cminyard@mvista.com>
Acked-by: Takashi Iwai <tiwai@suse.de>

---
 include/linux/ipmi.h     | 8 ++++++++
 include/linux/ipmi_smi.h | 6 ++++++
 2 files changed, 14 insertions(+)

diff --git a/include/linux/ipmi.h b/include/linux/ipmi.h
index 3474f04cf9aa..d89bea1e457a 100644
--- a/include/linux/ipmi.h
+++ b/include/linux/ipmi.h
@@ -97,6 +97,14 @@ struct ipmi_user_hndl {
 	 * been set up to handle run to completion.
 	 */
 	void (*ipmi_panic_handler)(void *handler_data);
+
+	/*
+	 * Called when the interface has been removed.  After this returns
+	 * the user handle will be invalid.  The interface may or may
+	 * not be usable when this is called, but it will return errors
+	 * if it is not usable.
+	 */
+	void (*shutdown)(void *handler_data);
 };
 
 /* Create a new user of the IPMI layer on the given interface number. */
diff --git a/include/linux/ipmi_smi.h b/include/linux/ipmi_smi.h
index 9e5c3079d232..bdcda4741c89 100644
--- a/include/linux/ipmi_smi.h
+++ b/include/linux/ipmi_smi.h
@@ -72,6 +72,12 @@ struct ipmi_smi_handlers {
 	int (*start_processing)(void       *send_info,
 				ipmi_smi_t new_intf);
 
+	/*
+	 * When called, the low-level interface should disable all
+	 * processing, it should be complete shut down when it returns.
+	 */
+	void (*shutdown)(void *send_info);
+
 	/*
 	 * Get the detailed private info of the low level interface and store
 	 * it into the structure of ipmi_smi_data. For example: the
-- 
2.19.2

